### PR TITLE
Add aggregate output metrics per input

### PR DESCRIPTION
**Pull request by AndersonQ**
## Proposed commit message

```
filebeat/input/v2: add output metrics aggregated by input

This change adds output metrics aggregated by input to the v2 inputs. It adds an
output listener passed to the outputs through the publisher.Event. The 
outputs.Observer interface was modified to track events individually, allowing
for output metrics to be aggregated by input. All the inputs were adjusted to 
use the new interface.
```

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- ~~[ ] I have made corresponding changes to the documentation~~
- ~~[ ] I have made corresponding change to the default configuration files~~
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Disruptive User Impact

- None. New metrics will be available, but they aren't considered GA and might change without previous notice

## Author's Checklist

- [x] generic output test: compare current metrics with new ones. Not a comprehensive test
  - [x] elasticsearch - it's comprehensive for ES as it uses mock-es
  - [x] kafka
  - [x] logstash
  - [x] redis
  - [x] otelconsumer
 - [x] update filebeat pipeline metrics per input integration test to also include output metrics

## Open question

Currently we don't have tests in the outputs for the metrics, especially for the edge/error cases. I've added some, for the happy path ensuring the new metrics align with the current metrics. But ideally more tests are needed.

I believe is worth discussing how much coverage we want and the strategy for testing, like having a mock for some outputs as we mock-es have for elasticsearch or trying to rely more on "unit" tests mocking the client instead of the output itself, just as I did for kafka.

### Option A: High-Fidelity Mock Servers
This approach would mirror our mock-es strategy by building lightweight, in-process mock servers for Kafka, Logstash, and Redis.

 - Pros:

    - Offers the highest fidelity by testing the full end-to-end output process.

    - More likely to catch subtle issues related to data framing or concurrency.

 - Cons:

    - Requires a significant upfront development effort.

    - The mock servers themselves will require ongoing maintenance.

### Option B: Unit-Level "Client" Mocks
This strategy relies on mocking at the client level, which is faster to implement.

#### Redis: Override the client's publish function to simulate success, retry, and drop scenarios.

 - Pros:

    - Quick to implement, fast, and deterministic.

 - Cons:

    - Doesn't test the full wire protocol, meaning some integration bugs could be missed.

    - Prevents black-box testing, meaning a generic test harness cannot be applied to all outputs.

### Path Forward: Option B
For the sake of velocity, we will adopt Option B for now. This allows us to add the necessary test coverage quickly.

However, it is worth considering a future investment in creating mock servers (per Option A) to create a more robust and generic testing framework for all outputs.

## How to test this PR locally

 - [Benchmark against latest release](https://ingest-metricstore.kb.us-central1.gcp.cloud.es.io:9243/app/r/s/WKAqZ)
if for some reason the link does not work, try filtering by: `environment_id: 019686f9-90a1-4da3-9ba7-d41ff730b73a` and time `Apr 30, 2025 @ 16:16:35.778 -> Apr 30, 2025 @ 17:16:35.778` 

The main test to ensure there is no changes in behaviour is to run filebeat with the same config on main and on this branch. The current global metrics should be the same on both branches. Once filebeat stops, you can compare the `Total metrics` log.

E.g: Running `TestInputMetricsFromPipelineAndOutput_Dropped`

 - main:
```
{"log.level":"info","@timestamp":"2025-06-04T07:35:23.790+0200","log.logger":"monitoring","log.origin":{"function":"github.com/elastic/beats/v7/libbeat/monitoring/report/log.(*reporter).logTotals","file.name":"log/log.go","file.line":200},"message":"Total metrics","service.name":"filebeat","monitoring":{"metrics":{"beat":{"cgroup":{"cpu":{"id":"vte-spawn-f1e2dbfd-9abc-45cf-9aca-cba911322588.scope","stats":{"periods":0,"throttled":{"ns":0,"periods":0}}},"memory":{"id":"vte-spawn-f1e2dbfd-9abc-45cf-9aca-cba911322588.scope","mem":{"usage":{"bytes":3656204288}}}},"cpu":{"system":{"ticks":30,"time":{"ms":30}},"total":{"ticks":80,"time":{"ms":80},"value":80},"user":{"ticks":50,"time":{"ms":50}}},"handles":{"limit":{"hard":1048576,"soft":1048575},"open":13},"info":{"ephemeral_id":"feb5366f-e1aa-4e55-964d-23f157f40f40","name":"filebeat","uptime":{"ms":1302},"version":"9.1.0"},"memstats":{"gc_next":49688194,"memory_alloc":24950312,"memory_sys":58545416,"memory_total":43320488,"rss":140382208},"runtime":{"goroutines":12}},"filebeat":{"harvester":{"closed":1,"open_files":0,"running":0,"skipped":0,"started":1},"input":{"log":{"files":{"renamed":0,"truncated":0}}}},"libbeat":{"config":{"module":{"running":0,"starts":0,"stops":0},"reloads":0,"scans":0},"output":{"batches":{"split":0},"events":{"acked":10,"active":0,"batches":3,"dead_letter":0,"dropped":1,"duplicates":1,"failed":1,"toomany":1,"total":13},"read":{"bytes":1954,"errors":0},"type":"elasticsearch","write":{"bytes":5994,"errors":0,"latency":{"histogram":{"count":3,"max":1,"mean":0.6666666666666666,"median":1,"min":0,"p75":1,"p95":1,"p99":1,"p999":1,"stddev":0.4714045207910317}}}}},"registrar":{"states":{"cleanup":0,"current":0,"update":0},"writes":{"fail":0,"success":0,"total":0}},"system":{"cpu":{"cores":16},"load":{"1":4.94,"15":3.8,"5":4.66,"norm":{"1":0.3088,"15":0.2375,"5":0.2913}}}},"ecs.version":"1.6.0"}}
```
```
"output":{"batches":{"split":0},"events":{"acked":10,"active":0,"batches":3,"dead_letter":0,"dropped":1,"duplicates":1,"failed":1,"toomany":1,"total":13}
```
 - this branch
```
{"log.level":"info","@timestamp":"2025-06-04T07:13:14.178+0200","log.logger":"monitoring","log.origin":{"function":"github.com/elastic/beats/v7/libbeat/monitoring/report/log.(*reporter).logTotals","file.name":"log/log.go","file.line":200},"message":"Total metrics","service.name":"filebeat","monitoring":{"metrics":{"beat":{"cgroup":{"cpu":{"id":"vte-spawn-7b861517-a94c-4efa-aa97-371697db16e2.scope","stats":{"periods":0,"throttled":{"ns":0,"periods":0}}},"memory":{"id":"vte-spawn-7b861517-a94c-4efa-aa97-371697db16e2.scope","mem":{"usage":{"bytes":1153634304}}}},"cpu":{"system":{"ticks":30,"time":{"ms":30}},"total":{"ticks":90,"time":{"ms":90},"value":90},"user":{"ticks":60,"time":{"ms":60}}},"handles":{"limit":{"hard":1048576,"soft":1048575},"open":13},"info":{"ephemeral_id":"55910960-8de9-4e1c-b0fe-438fd5b5305c","name":"filebeat","uptime":{"ms":1737},"version":"9.1.0"},"memstats":{"gc_next":52467490,"memory_alloc":28044592,"memory_sys":58545416,"memory_total":43437384,"rss":138260480},"runtime":{"goroutines":12}},"filebeat":{"harvester":{"closed":1,"open_files":0,"running":0,"skipped":0,"started":1},"input":{"log":{"files":{"renamed":0,"truncated":0}}}},"libbeat":{"config":{"module":{"running":0,"starts":0,"stops":0},"reloads":0,"scans":0},"output":{"batches":{"split":0},"events":{"acked":10,"active":0,"batches":3,"dead_letter":0,"dropped":1,"duplicates":1,"failed":1,"toomany":1,"total":13},"read":{"bytes":1954,"errors":0},"type":"elasticsearch","write":{"bytes":5984,"errors":0,"latency":{"histogram":{"count":3,"max":0,"mean":0,"median":0,"min":0,"p75":0,"p95":0,"p99":0,"p999":0,"stddev":0}}}}},"registrar":{"states":{"cleanup":0,"current":0,"update":0},"writes":{"fail":0,"success":0,"total":0}},"system":{"cpu":{"cores":16},"load":{"1":7.11,"15":3.14,"5":4.53,"norm":{"1":0.4444,"15":0.1963,"5":0.2831}}}},"ecs.version":"1.6.0"}}
```
```
"output":{"batches":{"split":0},"events":{"acked":10,"active":0,"batches":3,"dead_letter":0,"dropped":1,"duplicates":1,"failed":1,"toomany":1,"total":13}
```

### compare unit tests with main
You can run `libbeat/outputs/TestOutputsMetrics` on main and on this branch. The global output metrics should be the same. Thus, running the same test on `main` should report the same global metrics for the output as it does on this branch.

 - update mock-es: `go get -u github.com/elastic/mock-es@8c3b6053f9b6`

add the following tests files on `main` to add the necessary tests, then comment out everything that breaks and the `if t.Failed() { ... }` so the metrics will be printed on success.

 - [libbeat/outputs/outputs_test.go](https://github.com/AndersonQ/beats/blob/42762-dropped-event-per-input/libbeat/outputs/outputs_test.go) 
 - [libbeat/outputs/outputs_integration_test.go](https://github.com/AndersonQ/beats/blob/42762-dropped-event-per-input/libbeat/outputs/outputs_integration_test.go) 

the `metrics registry` should be the same for main and for this branch

```
go test -v -run ^TestOutputsMetrics ./libbeat/outputs
```

if you wanna run tests for the outputs that cannot be mocked, start the containers:
```
cd libbeat
mage docker:composeUp
go test -v -tags integration -run ^TestOutputsMetrics ./libbeat/outputs
```

Example output
```
=== RUN   TestOutputsMetrics
=== RUN   TestOutputsMetrics/elasticsearch
    outputs_test.go:212: metrics registry snapshot: 
        {"batches":{"split":0},"events":{"acked":1,"active":0,"batches":1,"dead_letter":0,"dropped":1,"duplicates":1,"failed":1,"toomany":1,"total":4},"read":{"bytes":516,"errors":0},"write":{"bytes":728,"errors":0,"latency":{"histogram":{"count":1,"max":1,"mean":1,"median":1,"min":1,"p75":1,"p95":1,"p99":1,"p999":1,"stddev":0}}}}
--- PASS: TestOutputsMetrics (0.00s)
    --- PASS: TestOutputsMetrics/elasticsearch (0.00s)
PASS
ok  	github.com/elastic/beats/v7/libbeat/outputs	0.304s
```

## Related issues

- Closes https://github.com/elastic/beats/issues/42762
- Relates https://github.com/elastic/beats/issues/42325

### CODE REVIEW

**Comment by pierrehilbert** [comment]
@AndersonQ looks like you should regenerate the NOTICE.txt content.

**Comment by AndersonQ** [comment]
> @AndersonQ looks like you should regenerate the NOTICE.txt content.

done. 
Wasn't this check supposed to be required?

**Comment by pierrehilbert** [comment]
> > @AndersonQ looks like you should regenerate the NOTICE.txt content.
> 
> done. Wasn't this check supposed to be required?

It is required, this is the reason why check-docs was red here.

**Comment by AndersonQ** [comment]
> > > @AndersonQ looks like you should regenerate the NOTICE.txt content.
> > 
> > 
> > done. Wasn't this check supposed to be required?
> 
> It is required, this is the reason why check-docs was red here.

I meant like the `buildkite/beats` that has a `Required` tag on it.
For example the `golangci-lint`, it isn't required, the merge isn't blocked if it fails. 
I don't have the permissions to see the branch protection rules, but I think this check isn't required.

**Comment by AndersonQ** [comment]
I updated the `Open question` section with 2 options a picked one

**Comment by AndersonQ** [comment]
@cmacknz, @faec, when you have some time, could review this?

File: libbeat/outputs/observer.go [hunk 1]
```diff
@@ -17,19 +17,25 @@
...
+	NewBatch([]publisher.Event) // report events in a new batch being processed
 
-	RetryableErrors(int)  // report number of events with retryable errors
-	PermanentErrors(int)  // report number of events dropped due to permanent errors
-	DuplicateEvents(int)  // report number of events detected as duplicates (e.g. on resends)
-	DeadLetterEvents(int) // report number of failed events ingested to dead letter index
-	AckedEvents(int)      // report number of acked events
-	ErrTooMany(int)       // report too many requests response
+	RetryableErrors([]publisher.Event)  // report event had with retryable errors
+	PermanentError(publisher.Event)     // report event has been dropped due to permanent errors
```
Comment (faec) [inline libbeat/outputs/observer.go:32]: I don't like the redundancy of `PermanentError` / `PermanentErrors` and `AckedEvent` / `AckedEvents`. I'd prefer the single-value version, which is what we talked about beforehand, though `PermanentError(...publisher.Event)` etc could be a reasonable compromise (and would be nice for others like `RetryableErrors` too since they naturally arise as both individual events and arrays).

File: libbeat/outputs/elasticsearch/client.go [hunk 2]
```diff
@@ -82,12 +82,12 @@ type clientSettings struct {
 }
 
 type bulkResultStats struct {
-	acked        int // number of events ACKed by Elasticsearch
```
Comment (faec) [inline libbeat/outputs/elasticsearch/client.go:85]: I think it makes more sense to remove this struct entirely. The only thing it does is aggregate and report the totals, and since we need to pass in the full events now, there's no point in building redundant lists of the different event categories when we can just call the observer directly on each one from `applyItemStatus`.

File: libbeat/outputs/elasticsearch/client.go [hunk 3]
```diff
@@ -82,12 +82,12 @@ type clientSettings struct {
...
-	duplicates   int // number of events failed with `create` due to ID already being indexed
-	fails        int // number of events with retryable failures.
-	nonIndexable int // number of events with permanent failures.
-	deadLetter   int // number of failed events ingested to the dead letter index.
-	tooMany      int // number of events receiving HTTP 429 Too Many Requests
+	acked        []publisher.Event // number of events ACKed by Elasticsearch
```
Comment (cmacknz) [inline libbeat/outputs/elasticsearch/client.go:85]: Have you quantified the potential memory impact of keeping the references to the events around longer?

**Review by cmacknz — COMMENTED** [review]
Seeing all the changes directly in the outputs makes me realize that this implementation won't work with Beats receivers, so the usefulness of this will be very short lived unless I am missing something.

With Beats receivers everything flows through the otelconsumer output, which we plan to remove eventually in https://github.com/elastic/beats/issues/44105.

We would need at minimum a totally different implementation for Beats receivers, that possibly requires changes to the core collector pipeline making it much harder to do. This makes me think we should pause or abandon this approach until we know how we can accommodate it in the collector.

File: filebeat/input/v2/input.go [hunk 4]
```diff
@@ -145,29 +183,77 @@ func PrepareInputMetrics(
...
 
-	return reg, pc, func() {
+	return reg, pipeConnector, func() {
 		// Unregister the metrics when the input finishes running.
 		defer inputmon.CancelMetricsRegistry(
 			inputID, name, parent, log)
 	}
 }
 
+type OutputListener struct {
```
Comment (faec) [inline filebeat/input/v2/input.go:220]: `OutputListener` -> `outputListener`, this implementation shouldn't be visible outside the input package

File: libbeat/beat/pipeline.go [hunk 5]
```diff
@@ -204,3 +295,71 @@ func (c *CombinedClientListener) DroppedOnPublish(event Event) {
 	c.A.DroppedOnPublish(event)
 	c.B.DroppedOnPublish(event)
 }
+
+// CombinedOutputListener combines two OutputListeners.
+type CombinedOutputListener struct {
```
Comment (faec) [inline libbeat/beat/pipeline.go:300]: I'd rather this be an internal helper (`combinedOutputListener`) in the Filebeat input package, unless/until it has a second caller.

File: libbeat/beat/pipeline.go [hunk 6]
```diff
@@ -171,6 +187,81 @@ const (
 	DropIfFull
 )
 
+// CountOutputListener is a simple implementation of OutputListener that uses
+// atomic counters. It's intended for use in tests.
+type CountOutputListener struct {
```
Comment (faec) [inline libbeat/beat/pipeline.go:192]: `CountOutputListener` and its call sites seem like they increase test complexity / API surface area without increasing coverage. The outputs themselves (and their tests) have no reason to know that OutputListener even exists. There's two things that need to be tested:
- That `outputs.Stats` calls through to the `OutputListener` of any events it is given. This can be verified in a unit test in the `outputs` package without exporting anything or linking to non-test builds, and it can use a minimal helper with public fields, no atomics, and no accessors or special helpers.
- That the outputs invoke the appropriate methods on their `outputs.Observer`. Their contract has nothing to do with any `OutputListener` the `Observer` might use internally, and can be verified by unit tests on the appropriate output that use the existing `outputs.Stats`. Many scenarios are already covered this way (e.g. see the earlier cases in `client_test.go:TestPublish` in the Elasticsearch output), and we can add more cases if needed, but I don't see any reason they would need special cross-package access to the internals of the observer.

File: libbeat/publisher/event.go [hunk 7]
```diff
@@ -60,13 +60,72 @@ type Event struct {
...
 	// If the output provides an early encoder for incoming events,
 	// it should store the encoded form in EncodedEvent and clear Content
 	// to free the unencoded data. The updated event will be provided to
 	// output workers when calling Publish.
 	EncodedEvent interface{}
 }
 
+// OutputListener is a helper around a beat.OutputListener to make it safe to
+// call methods on Event.OutputListener without requiring a nil check.
+type OutputListener struct {
```
Comment (faec) [inline libbeat/publisher/event.go:76]: I don't like making a special exported type (plus its own unit test) just to wrap these nil checks:
- The listener is already guaranteed not to be nil.
- If we want to do nil checks anyway (which seems reasonable), the only place they ever need to happen is in `outputs.Stats`, and the number of checks required would be exactly the same as the number added by this special wrapper object, so there's no need for a cross-package API.
- If there's some corner case in the tests where nil checks are tedious, they can just use `NoopOutputListener` as needed.

File: libbeat/publisher/pipeline/testing.go [hunk 8]
```diff
@@ -56,9 +56,9 @@ type mockNetworkClient struct {
 
 func (c *mockNetworkClient) Connect(_ context.Context) error { return nil }
 
-type mockBatch struct {
-	mu     sync.Mutex
-	events []publisher.Event
+type MockBatch struct {
```
Comment (faec) [inline libbeat/publisher/pipeline/testing.go:59]: Rather than expose this internal mocking helper, I'd suggest using `pipeline.NewBatchForTesting`. I don't think it makes sense to export a "mocked" batch implementation while leaving its actual callback hooks private, and since you don't need those hooks for your tests I think the real batch implementation would cover everything you need.

File: libbeat/outputs/outputs_test.go [hunk 9]
```diff
@@ -0,0 +1,290 @@
...
+
+type mockIndexSelector string
+
+func (m mockIndexSelector) Select(*beat.Event) (string, error) {
+	return string(m), nil
+}
+
+var _ publisher.Batch = (*mockBatch)(nil)
+
+type mockBatch struct {
```
Comment (faec) [inline libbeat/outputs/outputs_test.go:267]: Here I'd also suggest removing this mostly-empty implementation and just using `pipeline.NewBatchForTesting` if it suffices.

**Review by faec — COMMENTED** [review]
I'm still working my way through all the details but here's a first pass. The core structure seems good but I think there's several places where we can simplify the API / testing, use internal rather than public helpers, and remove unnecessary boilerplate.

File: libbeat/outputs/elasticsearch/client.go (hunk 3 above)
Comment (AndersonQ) [reply libbeat/outputs/elasticsearch/client.go:85]: Not really, but it isn't a long lived struct. I'll run benchbuilder for this branch and check is anything stands out

**Comment by pierrehilbert** [comment]
@AndersonQ following the weekly team meeting, I think we can close this PR right?

**Comment by AndersonQ** [comment]
> @AndersonQ following the weekly team meeting, I think we can close this PR right?

yeah, closing it
